src/pad.rs

```diff
@@ -0,0 +1,363 @@
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use untrusted;
+use digest;
+use rand;
+
+pub trait PaddingAlgorithm {
+    fn encode(&self, msg: &[u8], out: &mut [u8], rng: &rand::SecureRandom) -> Result<(), ()>;
+
+    fn verify(&self, msg: untrusted::Input, encoded: untrusted::Input) 
+              -> Result<(), ()>;
+}
+
+pub struct PKCS1v15Padding {
+    pub digest_alg: &'static digest::Algorithm,
+    pub digestinfo_prefix: &'static [u8],
+}
+
+impl PaddingAlgorithm for PKCS1v15Padding {
+    // Implement padding procedure per EMSA-PKCS1-v1_5,
+    // https://tools.ietf.org/html/rfc3447#section-9.2.
+    fn encode(&self, msg: &[u8], out: &mut [u8], _: &rand::SecureRandom) -> Result<(), ()> {
+        let digest_len =
+            self.digestinfo_prefix.len() + self.digest_alg.output_len;
+
+        // Require at least 8 bytes of padding. Since we disallow keys smaller
+        // than 2048 bits, this should never happen anyway.
+        debug_assert!(out.len() >= digest_len + 11);
+        let pad_len = out.len() - digest_len - 3;
+        out[0] = 0;
+        out[1] = 1;
+        for i in 0..pad_len {
+            out[2 + i] = 0xff;
+        }
+        out[2 + pad_len] = 0;
+
+        let (digest_prefix, digest_dst) = out[3 + pad_len..].split_at_mut(
+            self.digestinfo_prefix.len());
+        digest_prefix.copy_from_slice(self.digestinfo_prefix);
+        digest_dst.copy_from_slice(
+            digest::digest(self.digest_alg, msg).as_ref());
+        Ok(())
+    }
+
+    fn verify(&self, msg: untrusted::Input, encoded: untrusted::Input)
+              -> Result<(), ()> {
+        encoded.read_all((), |decoded| {
+            if try!(decoded.read_byte()) != 0 ||
+               try!(decoded.read_byte()) != 1 {
+                return Err(());
+            }
+
+            let mut ps_len = 0;
+            loop {
+                match try!(decoded.read_byte()) {
+                    0xff => { ps_len += 1; },
+                    0x00 => { break; },
+                    _ => { return Err(()); }
+                }
+            }
+            if ps_len < 8 {
+                return Err(());
+            }
+
+            let decoded_digestinfo_prefix =
+                try!(decoded.skip_and_get_input(
+                        self.digestinfo_prefix.len()));
+            if decoded_digestinfo_prefix != self.digestinfo_prefix {
+                return Err(());
+            }
+
+            let digest_alg = self.digest_alg;
+            let decoded_digest =
+                try!(decoded.skip_and_get_input(digest_alg.output_len));
+            let digest = digest::digest(digest_alg, msg.as_slice_less_safe());
+            if decoded_digest != digest.as_ref() {
+                return Err(());
+            }
+            Ok(())
+        })
+    }
+
+
+}
+
+// Maximum supported length of the salt;
+const MAX_SLEN: usize = 512 / 8;
```

Comment:
Should we spell this out as "MAX_SALT_LEN"? I realize "sLen" comes from the spec but I'm not sure it will be clear to readers. Same for `MAX_MLEN` below.